node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodePrioritizer.java

```diff
@@ -168,14 +171,38 @@ private NodeCandidate candidateFrom(Node node, boolean isSurplus) {
                                                                          currentClusterSize));
         }
         else {
-            return NodeCandidate.createStandalone(node, isSurplus, false);
+            return NodeCandidate.createStandalone(node, isSurplus, false, exclusiveSwitch);
         }
     }
 
-    private boolean isReplacement(int nofNodesInCluster, int nodeFailedNodes) {
-        if (nodeFailedNodes == 0) return false;
+    /** Returns whether given node is on an exclusive switch */
+    private boolean onExclusiveSwitch(Node node, Optional<Node> parent) {
+        Node host = parent.orElse(node);
+        return unusedSwitch(host, node.allocation()
+                                      .map(allocation -> allocation.membership().cluster().id())
+                                      .orElseGet(clusterSpec::id));
+    }
+
+    /** Returns whether switch of host is unused by any existing candidates for given cluster */
+    private boolean unusedSwitch(Node host, ClusterSpec.Id cluster) {
+        if (host.switchHostname().isEmpty()) return true;
+        String switchHostname = host.switchHostname().get();
+        for (var candidate : nodes) {
```

Comment:
When we will try to add new nodes in `addNewDockerNodesOn()`, wont this lead to us only considering the node on the first host that we read from ZK (or whatever order the hosts in `allNodes` are)?

Reply:
How? Each node added is checked against all the (parents of) candidates already added.

Reply:
Lets say we need to allocate 1 node, we have 2 hosts: `a` and `b`. `b` has massive skew, so allocation to `a` should be preferred. Both `a` and `b` are on the same switch, `s`.

Since we need to allocate 1 node, we start by creating a virtual reserved node on both `a` and `b`: `a1` and `b1`, as we create them, we check if `nodes` already has a node on `s` (initially, that is not the case). After we add the first of `a1` and `b1`, the second one will be deemed as not on exclusive switch, so skew is completely ignored.

Reply:
So given two hosts on the same switch, you want to prefer the candidate on the host with the least skew?

Reply:
The point is that while `a1` and `b1` are on the same switch, only 1 of them will be actually allocated anyway, so they cant be in conflict with each other

Reply:
> only 1 of them will be actually allocated anyway

The exclusive switch isn't a hard requirement though. If you ask for 2 nodes, you will allocate both `a1` and `b1` if no better candidate on an exclusive switch exists.